src/clib/calc_all_tdust_gasgr_1d_g.F

```diff
@@ -0,0 +1,353 @@
+
+!  Arguments
+
+      integer in, jn, kn, is, ie, j, k, nratec, idustfield
+
+      real*8 fgr
+      real*8 gasgra(nratec), gamma_isrfa
+
+!  Parameters
+
+      real*8, parameter :: mh = mass_h      !DPC
+
+!  Locals
+
+      integer i
+      real*8 trad, coolunit
+
+!  Slice locals
+ 
+      integer*8 indixe(in)
+      real*8 tdef(in), tgas(in), tdust(in), nh(in), metallicity(in),
+     &       dust2gas(in)
+
+!  Cooling/heating slice locals
+
+      real*8 gasgr(in), gasgr_tdust(in), myisrf(in)
+      integer idspecies, itdmulti
+!     opacity table
+      integer gr_N(2), gr_Size
+      real*8  gr_dT, gr_Td(gr_N(2))
+!     grain growth
+      real*8  sgSiM(in), sgFeM(in), sgMg2SiO4(in)
+     &      , sgMgSiO3(in), sgFe3O4(in), sgAC(in)
+     &      , sgSiO2D(in), sgMgO(in), sgFeS(in)
+     &      , sgAl2O3(in)
+     &      , sgreforg(in), sgvolorg(in), sgH2Oice(in)
+     &      , sgtot(in)
+      real*8  alSiM(gr_N(2),in), alFeM(gr_N(2),in)
+     &      , alMg2SiO4(gr_N(2),in), alMgSiO3(gr_N(2),in)
+     &      , alFe3O4(gr_N(2),in), alAC(gr_N(2),in)
+     &      , alSiO2D(gr_N(2),in), alMgO(gr_N(2),in)
+     &      , alFeS(gr_N(2),in), alAl2O3(gr_N(2),in)
+     &      , alreforg(gr_N(2),in)
+     &      , alvolorg(gr_N(2),in), alH2Oice(gr_N(2),in)
+     &      , altot(gr_N(2),in)
+      real*8  kpSiM(in), kpFeM(in), kpMg2SiO4(in)
+     &      , kpMgSiO3(in), kpFe3O4(in), kpAC(in)
+     &      , kpSiO2D(in), kpMgO(in), kpFeS(in)
+     &      , kpAl2O3(in)
+     &      , kpreforg(in), kpvolorg(in), kpH2Oice(in)
+     &      , kptot(in)
+!     grain temperature
+      real*8  tSiM(in), tFeM(in), tMg2SiO4(in)
+     &      , tMgSiO3(in), tFe3O4(in), tAC(in)
+     &      , tSiO2D(in), tMgO(in), tFeS(in)
+     &      , tAl2O3(in)
+     &      , treforg(in), tvolorg(in), tH2Oice(in)
+      real*8  gasgr2a(nratec), gamma_isrf2a
+      real*8  gasSiM(in), gasFeM(in), gasMg2SiO4(in)
+     &      , gasMgSiO3(in), gasFe3O4(in), gasAC(in)
+     &      , gasSiO2D(in), gasMgO(in), gasFeS(in)
+     &      , gasAl2O3(in)
+     &      , gasreforg(in), gasvolorg(in), gasH2Oice(in)
+      real*8  gasgr_tSiM(in), gasgr_tFeM(in), gasgr_tMg2SiO4(in)
+     &      , gasgr_tMgSiO3(in), gasgr_tFe3O4(in), gasgr_tAC(in)
+     &      , gasgr_tSiO2D(in), gasgr_tMgO(in), gasgr_tFeS(in)
+     &      , gasgr_tAl2O3(in)
+     &      , gasgr_treforg(in), gasgr_tvolorg(in), gasgr_tH2Oice(in)
+      real*8  mygisrf(in), fv2k, fac
+      real*8  gisrfSiM(in), gisrfFeM(in), gisrfMg2SiO4(in)
+     &      , gisrfMgSiO3(in), gisrfFe3O4(in), gisrfAC(in)
+     &      , gisrfSiO2D(in), gisrfMgO(in), gisrfFeS(in)
+     &      , gisrfAl2O3(in)
+     &      , gisrfreforg(in), gisrfvolorg(in), gisrfH2Oice(in)
+!  Iteration mask
+
+      MASK_TYPE itmask_metal(in)
+
+          ! Calculate heating from interstellar radiation field
+          !  -> this is ONLY used when `itmask_metal .eq. MASK_TRUE`
+
+      do i = is+1, ie+1
+         if ( itmask_metal(i) .ne. MASK_FALSE ) then
```

Comment:
Out of curiosity, do you know if it's more performant to have one do loop and itmask check and then all of the parameter-based if statements inside or multiple do/itmask loops? Maybe this is something you've thought about for the transcription. I see this is code is basically copied from its original location, so no need to change it, just curious.

Reply:
Honestly, the answer is "it depends." As a general rule-of-thumb, you want to have as few branches within a for-loop as possible and generally, an if-statement corresponds to a branch (there are of course times when we can get clever and replace if-statements with a mathematical expression that doesn't have any branching).

Right now, it would definitely be faster to break this up into many for-loops. With that said, I actually think we could significantly refactor this function and get rid of a bunch of temporary buffers, which would be even better. I think it will all become a little more clear once its in C++.